fri/src/prover.rs

```diff
@@ -95,20 +96,57 @@ where
     let mut data = vec![];
 
     while folded.len() > config.blowup() * config.final_poly_len() {
-        let leaves = RowMajorMatrix::new(folded, 2);
-        let (commit, prover_data) = config.mmcs.commit_matrix(leaves);
+        let arity = config.arity();
+        let next_folded_len = folded.len() / arity;
+
+        // First, we collect the polynomial evaluations that will be committed this round.
+        // Those are `folded` and polynomials in `inputs` not consumed yet with number of
+        // evaluations more than `next_folded_len`
+        let mut polys_before_next_round = vec![];
+
+        let mut cur_folded_len = folded.len();
+        while cur_folded_len > next_folded_len {
+            if let Some(poly_eval) = inputs_iter.next_if(|v| v.len() == cur_folded_len) {
+                let width = poly_eval.len() / next_folded_len;
+                let poly_eval_matrix = RowMajorMatrix::new(poly_eval, width);
+                polys_before_next_round.push(poly_eval_matrix);
+            }
+
+            cur_folded_len /= 2;
+        }
+
+        let folded_matrix = RowMajorMatrix::new(folded.clone(), arity);
```

Comment:
it would be best to avoid this `folded.clone()` since that's an entire matrix. below, can you use `leaves` in place of `folded`? it would make the code a little more readable too

Reply:
it actually makes the code uglier since `leaves` gives you only a reference, so the first fold will be a bit special. I'll also need to figure out how to pass that matrix reference to `g.fold_matrix` but with the width changed (to 2). do you know how to do that?

Reply:
at least for now, `DenseMatrix` has `pub width` so you could just do `folded_matrix.width = 2` since the underlying storage is row major vector.

Reply:
`leaves` only give us an immutable reference

Reply:
noting we're leaving for now, but could probably use `.as_view()`